daemon/graphdriver/overlay2/check.go

```diff
@@ -124,6 +124,61 @@ func doesSupportNativeDiff(d string) error {
 	return nil
 }
 
+// Forked from https://github.com/containers/storage/blob/4d1ae4a7983ded01f2dbb97792884642c8363e36/drivers/overlay/check.go#L187
+//
+// doesSupportVolatile checks if the filesystem supports the "volatile" mount option
+func doesSupportVolatile(d string) (bool, error) {
+	userxattr := false
+	if userns.RunningInUserNS() {
+		needed, err := overlayutils.NeedsUserXAttr(d)
+		if err != nil {
+			return false, err
+		}
+		if needed {
+			userxattr = true
+		}
+	}
+
+	td, err := os.MkdirTemp(d, "volatile-check")
+	if err != nil {
+		return false, err
+	}
+	defer func() {
+		if err := os.RemoveAll(td); err != nil {
+			logger.WithError(err).Warnf("failed to remove check directory %v", td)
+		}
+	}()
+
+	lower, upper, work, merged := filepath.Join(td, "lower"), filepath.Join(td, "upper"), filepath.Join(td, "work"), filepath.Join(td, "merged")
+	for _, dir := range []string{lower, upper, work, merged} {
+		if err := os.Mkdir(dir, 0o755); err != nil {
+			return false, err
+		}
+	}
+
+	opts := []string{fmt.Sprintf("volatile,lowerdir=%s,upperdir=%s,workdir=%s", lower, upper, work)}
+	if userxattr {
+		opts = append(opts, "userxattr")
+	}
+
+	m := mount.Mount{
+		Type:    "overlay",
+		Source:  "overlay",
+		Options: opts,
+	}
+
+	if err := m.Mount(merged); err != nil {
+		return false, errors.Wrap(err, "failed to mount overlay for volatile check")
+	}
+	defer func() {
+		if err := mount.UnmountAll(merged, 0); err != nil {
+			logger.WithError(err).Warnf("failed to unmount check directory %v", merged)
+		}
+	}()
+
+	return true, nil
```

Comment:
Should we also have a post-mount verification mechanism like we with `metacopy` for example?

https://github.com/moby/moby/blob/62244ebdd2005b3e2adb9031b6b42fb4dcf18c08/daemon/graphdriver/overlay2/check.go#L258

Would in this case the `volatile` xattr also be present?
The overlayfs [docs](https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html#volatile-mount) mention the creation of `$workdir/work/incompat/volatile`, so perhaps we could use it instead?

Reply:
Seems reasonable to me, please see https://github.com/moby/moby/pull/50151/commits/71681f1fa0eb1505ab362dcf995be9837fa3109d

I also noticed a build failure due to incorrect `%q` usage, and I figure to provide `Using volatile` in the `docker info` output (since `Supports` != `Using` depending on the driver configuration).